app/controllers/concerns/signature_verification.rb

```diff
@@ -123,13 +136,36 @@ def build_signed_string(signed_headers)
   end
 
   def matches_time_window?
-    begin
-      time_sent = Time.httpdate(request.headers['Date'])
-    rescue ArgumentError
-      return false
+    created_time = nil
+    expires_time = nil
+
+    if signature_algorithm == 'hs2019' && signature_params['created'].present?
+      created_time = Time.at(signature_params['created'].to_i).utc
+    elsif request.headers['Date'].present?
+      begin
+        created_time = Time.httpdate(request.headers['Date']).utc
+      rescue ArgumentError
+        return false
+      end
     end
 
-    (Time.now.utc - time_sent).abs <= 12.hours
+    expires_time = Time.at(signature_params['expires'].to_i).utc if signature_params['expires'].present?
+    expires_time ||= created_time + 5.minutes unless created_time.nil?
+
+    # It is possible for a legitimate user to sign `(created)` but not `(expires)`
+    # therefore, an attacker could set a long expiration time, so set a limit to
+    # how far in the future we allow the expiration to be.
+    expires_time = [expires_time, created_time + 12.hours].min unless created_time.nil?
```

Comment:
let's add a constant to replace this magic `12.hours`—this should make the code more readable and allow us to get rid of the comment.

Reply:
something like `EXPIRATION_WINDOW_LIMIT = 12.hours`

Reply:
Also, maybe we should only process `expires` if it's part of the signed_headers to avoid allowing the attacker to control this? I can't see how it would be harmful to allow the attacker to control this param, but seems good to do as a matter of best practices anyway. Either way we should limit the signature to at most 12 hours.

Reply:
Addressed the first two comments. Not sure about checking whether `expires` is within the `signed_headers`. I guess with the current implementation, what an attacker could do with it is extend the 1h5min (including clock skew margin) window to 13h (including clock skew margin)… I'm not sure that's worth making the code more complex. The 1h5 minutes is not something mandated by the spec in any way either, the spec doesn't tell what to do if there's no `expires` attribute.